pkg/frontend/v2/frontend.go

```diff
@@ -308,14 +310,30 @@ func (f *Frontend) CheckReady(_ context.Context) error {
 type requestsInProgress struct {
 	mu       sync.Mutex
 	requests map[uint64]*frontendRequest
+	logger   log.Logger
+	wg       sync.WaitGroup
 }
 
-func newRequestsInProgress() *requestsInProgress {
+func newRequestsInProgress(logger log.Logger) *requestsInProgress {
 	return &requestsInProgress{
 		requests: map[uint64]*frontendRequest{},
+		logger:   logger,
 	}
 }
 
+// Wait waits for any in-flight requests to finish.
+func (r *requestsInProgress) Wait() {
+	if r.count() == 0 {
+		level.Info(r.logger).Log("msg", "ready to terminate, no in-flight queries remain")
```

Comment:
The log says "ready to terminate" but `requestsInProgress` should be agnostic of how the caller uses it. I would keep it simple and just remove this whole `if` branch` and always run the following code.

Reply:
@pracucci sounds good, will do.

Reply:
Done.